### ISSUE TITLE
Print method for tbl_sqlserver giving errors

### ISSUE THREAD

**Issue by Hong-Revo** [imanuelcostigan/RSQLServer#142]
Not sure what's causing it yet.

```
library(DBI)
library(RSQLServer)
library(dplyr)

db <- src_sqlserver("host", database=*, properties=list(user=*, password=*)

indata <- tbl(db, "mytable")
indata
#Source:     table<mytable> [?? x 18]
#Database:   SQLServer 13.0.4001 [...]
#
#Error: length(order_by) not greater than 0L
```
Traceback() gives:
```
25: stop(assertError(attr(res, "msg")))
24: assertthat::assert_that(length(order_by) > 0L, dbGetInfo(con)$db.version >= 
        11, is.integer(offset), offset >= 0)
23: sql_select.SQLServerConnection(con, query$select, from, where = query$where, 
        group_by = query$group_by, having = query$having, order_by = query$order_by, 
        limit = query$limit, distinct = query$distinct, ...)
22: sql_select(con, query$select, from, where = query$where, group_by = query$group_by, 
        having = query$having, order_by = query$order_by, limit = query$limit, 
        distinct = query$distinct, ...)
21: sql_render.select_query(query$from, con, ..., root = root)
20: sql_render(query$from, con, ..., root = root)
19: inherits(x, "ident")
18: is.ident(from)
17: sql_subquery.default(con, sql_render(query$from, con, ..., root = root), 
        name = NULL)
16: sql_subquery(con, sql_render(query$from, con, ..., root = root), 
        name = NULL)
15: sql_render.select_query(qry, con = con, ...)
14: sql_render(qry, con = con, ...)
13: sql_render.tbl_lazy(x, con, con = <S4 object of class "SQLServerConnection">)
12: NextMethod(con = con)
11: sql_render.tbl_sql(x, con)
10: sql_render(x, con)
9: collect.tbl_sql(x, n = n)
8: collect(x, n = n)
7: as.data.frame(collect(x, n = n))
6: as.data.frame.tbl_sql(head(x, n))
5: as.data.frame(head(x, n))
4: trunc_mat(x, n = n, width = width)
3: print(trunc_mat(x, n = n, width = width))
2: print.tbl_sql(x)
1: function (x, ...) 
   UseMethod("print")(x)
```

**Comment by imanuelcostigan** [comment]
Strange. I can't reproduce. 

NB: I do not attach RSQLServer to search path (it is after all only a backend for DBI and dplyr interfaces. But this doesn't introduce an error for me

Using test server:

``` r
library(DBI);library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
ss <- RSQLServer::src_sqlserver("TEST", database = "DBItest")
tbl(ss, "airlines")
#> Source:   query [?? x 2]
#> Database: SQLServer 13.0.1722 [sqluser@13.70.190.239:1433/DBItest]
#> 
#>    carrier                        name
#>      <chr>                       <chr>
#> 1       9E           Endeavor Air Inc.
#> 2       AA      American Airlines Inc.
#> 3       AS        Alaska Airlines Inc.
#> 4       B6             JetBlue Airways
#> 5       DL        Delta Air Lines Inc.
#> 6       EV    ExpressJet Airlines Inc.
#> 7       F9      Frontier Airlines Inc.
#> 8       FL AirTran Airways Corporation
#> 9       HA      Hawaiian Airlines Inc.
#> 10      MQ                   Envoy Air
#> # ... with more rows
devtools::session_info()
#> Session info --------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.3.2 (2016-10-31)
#>  system   x86_64, darwin16.3.0        
#>  ui       unknown                     
#>  language (EN)                        
#>  collate  en_AU.UTF-8                 
#>  tz       Australia/Sydney            
#>  date     2017-03-04
#> Packages ------------------------------------------------------------------
#>  package    * version    date      
#>  assertthat   0.1        2013-12-06
#>  backports    1.0.5      2017-01-18
#>  DBI        * 0.5-19     2017-03-04
#>  devtools     1.12.0     2016-06-24
#>  digest       0.6.12     2017-01-27
#>  dplyr      * 0.5.0.9000 2017-03-04
#>  evaluate     0.10       2016-10-11
#>  htmltools    0.3.5      2016-03-21
#>  knitr        1.15.1     2016-11-22
#>  magrittr     1.5        2014-11-22
#>  memoise      1.0.0      2016-01-29
#>  purrr        0.2.2      2016-06-18
#>  R6           2.2.0      2016-10-05
#>  Rcpp         0.12.9     2017-01-14
#>  rJava        0.9-8      2016-01-07
#>  rmarkdown    1.3        2016-12-21
#>  rprojroot    1.2        2017-01-16
#>  RSQLServer   0.2.099    2017-03-04
#>  stringi      1.1.2      2016-10-01
#>  stringr      1.2.0      2017-02-18
#>  tibble       1.2        2016-08-26
#>  withr        1.0.2      2016-06-20
#>  yaml         2.1.14     2016-11-12
#>  source                                    
#>  CRAN (R 3.3.0)                            
#>  CRAN (R 3.3.2)                            
#>  Github (rstats-db/DBI@d633b79)            
#>  CRAN (R 3.3.1)                            
#>  CRAN (R 3.3.2)                            
#>  Github (hadley/dplyr@63d4a9f)             
#>  CRAN (R 3.3.1)                            
#>  CRAN (R 3.3.0)                            
#>  CRAN (R 3.3.2)                            
#>  CRAN (R 3.3.0)                            
#>  CRAN (R 3.3.0)                            
#>  CRAN (R 3.3.1)                            
#>  CRAN (R 3.3.1)                            
#>  CRAN (R 3.3.2)                            
#>  CRAN (R 3.3.0)                            
#>  CRAN (R 3.3.2)                            
#>  CRAN (R 3.3.2)                            
#>  local (imanuelcostigan/RSQLServer@ec9226f)
#>  CRAN (R 3.3.1)                            
#>  CRAN (R 3.3.2)                            
#>  CRAN (R 3.3.1)                            
#>  CRAN (R 3.3.0)                            
#>  CRAN (R 3.3.2)
```

**Comment by Hong-Revo** [comment]
I can repro on a local SQL 2016 instance, as well as a remote instance in Azure. Looks like you've found the cause though?

**Comment by Hong-Revo** [comment]
Commit at https://github.com/hadley/dplyr/commit/ae202a14e76a86c8bd3cef58119f81b94395a70e has broken `collect` when no. of rows is specified.

**Comment by imanuelcostigan** [comment]
`collect()` works sometimes and doesn't work other times:
``` r
# Using https://github.com/imanuelcostigan/RSQLServer/commit/93875689373ddd57da2adeb90c7d06d577c58843
library(DBI); library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
ss <- RSQLServer::src_sqlserver("TEST", database = "DBItest")
df1 <- data_frame(a = c(1, 2), b = c("a", "b"))
copy_to(ss, df1, temporary = FALSE)
#> Error in rJava::.jcall(statement, "I", "executeUpdate", string, check = check): java.sql.SQLException: There is already an object named 'df1' in the database.
df1_tbl <- tbl(ss, "df1")
# Works
df1_tbl %>% collect()
#> # A tibble: 2 × 2
#>       a     b
#> * <dbl> <chr>
#> 1     1     a
#> 2     2     b
# Doesn't work
df1_tbl %>% select(a) %>% collect()
#> Error: length(order_by) not greater than 0L
```

**Comment by Hong-Revo** [comment]
Yes, it's fine if you want all the rows. It's only if you specify the number of rows to collect that it breaks.

This includes the situation where you run a chain without assigning the result. This is because the result gets printed to the screen, and the `print` method for tbls prints the first few rows, which means calling `collect` with rows specified.

**Comment by imanuelcostigan** [comment]
Note this isn't an issue for SQLite:

``` r
``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(nycflights13)
flights_db1 <- tbl(nycflights13_sqlite(), "flights")
#> Caching nycflights db at /Users/imanuel/RLibrary/dplyr/db/nycflights13.sqlite
flights_db1 %>% collect()
#> # A tibble: 336,776 × 19
#>     year month   day dep_time sched_dep_time dep_delay arr_time
#>    <int> <int> <int>    <int>          <int>     <dbl>    <int>
#> 1   2013     1     1      517            515         2      830
#> 2   2013     1     1      533            529         4      850
#> 3   2013     1     1      542            540         2      923
#> 4   2013     1     1      544            545        -1     1004
#> 5   2013     1     1      554            600        -6      812
#> 6   2013     1     1      554            558        -4      740
#> 7   2013     1     1      555            600        -5      913
#> 8   2013     1     1      557            600        -3      709
#> 9   2013     1     1      557            600        -3      838
#> 10  2013     1     1      558            600        -2      753
#> # ... with 336,766 more rows, and 12 more variables: sched_arr_time <int>,
#> #   arr_delay <dbl>, carrier <chr>, flight <int>, tailnum <chr>,
#> #   origin <chr>, dest <chr>, air_time <dbl>, distance <dbl>, hour <dbl>,
#> #   minute <dbl>, time_hour <dbl>
flights_db1 %>% select(year) %>% collect()
#> # A tibble: 336,776 × 1
#>     year
#>    <int>
#> 1   2013
#> 2   2013
#> 3   2013
#> 4   2013
#> 5   2013
#> 6   2013
#> 7   2013
#> 8   2013
#> 9   2013
#> 10  2013
#> # ... with 336,766 more rows
devtools::session_info()
#> Session info --------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.3.2 (2016-10-31)
#>  system   x86_64, darwin16.3.0        
#>  ui       unknown                     
#>  language (EN)                        
#>  collate  en_AU.UTF-8                 
#>  tz       Australia/Sydney            
#>  date     2017-03-05
#> Packages ------------------------------------------------------------------
#>  package      * version    date       source                        
#>  assertthat     0.1        2013-12-06 CRAN (R 3.3.0)                
#>  backports      1.0.5      2017-01-18 CRAN (R 3.3.2)                
#>  DBI            0.5-19     2017-03-04 Github (rstats-db/DBI@c018ed3)
#>  devtools       1.12.0     2016-06-24 CRAN (R 3.3.1)                
#>  digest         0.6.12     2017-01-27 CRAN (R 3.3.2)                
#>  dplyr        * 0.5.0.9000 2017-03-04 Github (hadley/dplyr@04df574) 
#>  evaluate       0.10       2016-10-11 CRAN (R 3.3.1)                
#>  htmltools      0.3.5      2016-03-21 CRAN (R 3.3.0)                
#>  knitr          1.15.1     2016-11-22 CRAN (R 3.3.2)                
#>  lazyeval       0.2.0      2016-06-12 cran (@0.2.0)                 
#>  magrittr       1.5        2014-11-22 CRAN (R 3.3.0)                
#>  memoise        1.0.0      2016-01-29 CRAN (R 3.3.0)                
#>  nycflights13 * 0.2.2      2017-01-27 CRAN (R 3.3.2)                
#>  R6             2.2.0      2016-10-05 CRAN (R 3.3.1)                
#>  Rcpp           0.12.9     2017-01-14 CRAN (R 3.3.2)                
#>  rmarkdown      1.3        2016-12-21 CRAN (R 3.3.2)                
#>  rprojroot      1.2        2017-01-16 CRAN (R 3.3.2)                
#>  RSQLite        1.1-2      2017-01-08 CRAN (R 3.3.2)                
#>  stringi        1.1.2      2016-10-01 CRAN (R 3.3.1)                
#>  stringr        1.2.0      2017-02-18 CRAN (R 3.3.2)                
#>  tibble         1.2        2016-08-26 CRAN (R 3.3.1)                
#>  withr          1.0.2      2016-06-20 CRAN (R 3.3.0)                
#>  yaml           2.1.14     2016-11-12 CRAN (R 3.3.2)
```

**Comment by imanuelcostigan** [comment]
Here is the call stack on error:

```
20: stop(assertError(attr(res, "msg")))
19: assertthat::assert_that(length(order_by) > 0L, dbGetInfo(con)$db.version >= 
        11, is.integer(offset), offset >= 0)
18: sql_select.SQLServerConnection(con, query$select, from, where = query$where, 
        group_by = query$group_by, having = query$having, order_by = query$order_by, 
     ...
17: sql_select(con, query$select, from, where = query$where, group_by = query$group_by, 
        having = query$having, order_by = query$order_by, limit = query$limit, 
     ...
16: sql_render.select_query(qry, con = con, ...)
15: sql_render(qry, con = con, ...)
14: sql_render.tbl_lazy(x, con, con = <S4 object of class "SQLServerConnection">)
13: NextMethod(con = con)
12: sql_render.tbl_sql(x, con)
11: sql_render(x, con)
10: collect.tbl_sql(.)
9: collect(.)
8: function_list[[k]](value)
7: withVisible(function_list[[k]](value))
6: freduce(value, `_function_list`)
5: `_fseq`(`_lhs`)
4: eval(expr, envir, enclos)
3: eval(quote(`_fseq`(`_lhs`)), env, env)
2: withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
1: df1_tbl %>% select(a) %>% collect()
```

**Comment by Hong-Revo** [comment]
See my comment on the dplyr commit. If a database uses `sql_select.default` there is no problem. `sql_select.SQLServerConnection` trips the bug because it has extra arguments, which will catch anything passed down via `...`

**Comment by imanuelcostigan** [comment]
When debugging in step 12, here is what I see as arguments in call to `sql_render.tbl_sql()`:

* `query` has following class: `"tbl_sqlserver" "tbl_sql"       "tbl_lazy"      "tbl"`
* `con` has following class: `"SQLServerConnection"`
* `list(...)` has value: `list()`

When debugging in step 14 (after call to `NextMethod(con = con)`, the arguments in call to `sql_render.tbl_lazy()`: 

* `query` has following class: `"tbl_sqlserver" "tbl_sql"       "tbl_lazy"      "tbl"`
* `con` has following class: `"SQLServerConnection"`
* `list(...)` has value: `[[1]] <SQLServerConnection>`

In this call, `qry <- sql_build(query, con = con, ...)` returns following value for  `qry`:

```
List of 8
 $ from    :Classes 'ident', 'sql', 'character'  chr "df1"
 $ select  :Classes 'ident', 'sql', 'character'  Named chr "a"
  .. ..- attr(*, "names")= chr "a"
 $ where   : chr(0) 
 $ group_by: chr(0) 
 $ having  : chr(0) 
 $ order_by: chr(0) 
 $ distinct: logi FALSE
 $ limit   : NULL
 - attr(*, "class")= chr [1:2] "select_query" "query"
```

Then in call to `sql_render.select_query()`, `...`, which contains a `SQLServerConnection` object, is passed through to `sql_select.SQLServerConnection`. As the argument list of this `sql_select` method has arguments past the standard ones from dplyr (starting with the additional argument `offset`), the `SQLServerConnection` object is passed through to the `offset` argument. This causes code to execute erroneously in `sql_select.SQLServerConnection`

**Comment by imanuelcostigan** [comment]
closed by #145
